### PR TITLE
Add bicubic interpolation support for affine transforms (resolves #351)

### LINKED ISSUE #351: Add bicubic interpolation
Currently the affine and rotation functions only support nearest neighbor or bilinear interpolation.  It would be nice to add bicubic interpolation support for slightly nicer transform outputs.

### PR DESCRIPTION
**Pull request by mduggan**
I have implemented a version of this.  Some caveats:
- It's my first attempt at really doing anything in Rust so let me know where my code sucks :)
- It loses a pixel off the edge because of the 4x4 kernel. I want to improve edge behavior for all the affine code, but that would be a separate change.
- It's not super fast.. about 1/2 the speed of the bilinear stuff.  Here are the bench results on my laptop:
```
test affine::tests::bench_affine_bicubic                                       ... bench:   2,054,455 ns/iter (+/- 574,716)
test affine::tests::bench_affine_bilinear                                      ... bench:     822,706 ns/iter (+/- 198,371)
test affine::tests::bench_affine_nearest                                       ... bench:     390,691 ns/iter (+/- 22,217)
test affine::tests::bench_rotate_bicubic                                       ... bench:     804,275 ns/iter (+/- 182,569)
test affine::tests::bench_rotate_bilinear                                      ... bench:     430,449 ns/iter (+/- 23,885)
test affine::tests::bench_rotate_nearest                                       ... bench:     292,266 ns/iter (+/- 26,284)
```

### CODE REVIEW

**Comment by theotherphil** [comment]
This looks good, thanks!

Unfortunately it’s going to have a massive merge conflict with this other outstanding PR: https://github.com/image-rs/imageproc/pull/350

I’m planning to merge that one first. Your change can then hopefully be added on top of it.

**Comment by mduggan** [comment]
Ah, thanks for letting me know. It won't merge automatically but it doesn't look too hard to manually move my changes over - I will make a new version based on top of that branch.

**Comment by theotherphil** [comment]
I'll merge https://github.com/image-rs/imageproc/pull/350 soon, and then tinker with the formatting a bit and extend the docs. I'll let you know when I'm done so that you can add your changes without risk of the file changing underneath you.

**Comment by theotherphil** [comment]
I'm now not going to make any more changes to geometric_transformations until the changes in this PR have been merged.

**Comment by mduggan** [comment]
Thanks, updated and I think it should be good now.

File: src/geometric_transformations.rs [hunk 1]
```diff
@@ -568,7 +574,70 @@ fn mul3x3(a: [f32; 9], b: [f32; 9]) -> [f32; 9] {
...
+    P::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let mut outp = px0.clone();
+
+    for i in 0..(P::channel_count() as usize) {
+        let p0 = cast(px0.channels()[i]);
+        let p1 = cast(px1.channels()[i]);
+        let p2 = cast(px2.channels()[i]);
+        let p3 = cast(px3.channels()[i]);
+        let pval = p1 + 0.5 * x * (p2 - p0 + x * (2.0*p0 - 5.0*p1 + 4.0*p2 - p3 + x*(3.0*(p1 - p2) + p3 - p0)));
```
Comment (theotherphil) [inline src/geometric_transformations.rs]: Nit: please add spaces around all operators here and in interpolate_bicubic.

**Comment by theotherphil** [comment]
I’ve left one small nit, but otherwise LGTM.

**Comment by mduggan** [comment]
Fixed the whitespace, happy to add some tests - will try to find the time to do that today.

**Comment by mduggan** [comment]
Done.

**Comment by theotherphil** [comment]
Thanks!

**Pull request merged**